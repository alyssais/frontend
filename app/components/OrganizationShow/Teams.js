// @flow

import * as React from 'react';
import { createFragmentContainer, graphql } from 'react-relay/compat';
import Dropdown from 'app/components/shared/Dropdown';
import Chooser from 'app/components/shared/Chooser';
import Emojify from 'app/components/shared/Emojify';
import Icon from 'app/components/shared/Icon';
import type { Teams_organization } from './__generated__/Teams_organization.graphql';

type Props = {
  selected: string,
  organization: Teams_organization,
  onTeamChange: (teamSlug: string) => void
};

class Teams extends React.Component<Props> {
  dropdownNode: ?Dropdown;

  get teamEdges() {
    if (this.props.organization.teams && this.props.organization.teams.edges) {
      return this.props.organization.teams.edges;
    }
    return [];
  }

  render() {
    // Collect all the teams that we're allowed to see pipelines on
    const teams = this.teamEdges.reduce((teams, teamEdge) => (
      teamEdge &&
      teamEdge.node &&
      teamEdge.node.permissions &&
      teamEdge.node.permissions.pipelineView &&
      teamEdge.node.permissions.pipelineView.allowed ? (
          teams.concat(teamEdge.node)
        ) : (
          teams
        )
    ), []);


    // Don't render the select if there aren't any teams that can be viewed
    if (teams.length === 0) {
      return null;
    }

    return (
      <Dropdown className="ml4" width={300} ref={(dropdownNode) => this.dropdownNode = dropdownNode}>
        <button className="h3 px0 py1 m0 light dark-gray inline-block btn" style={{ fontSize: 16 }}>
          <div className="flex">
            <span className="flex items-center">
              <span className="truncate">
                {this.renderLabel()}
              </span>
            </span>
            <span className="flex items-center">
              <Icon icon="down-triangle" style={{ width: 8, height: 8, marginLeft: '.5em' }} />
            </span>
          </div>
        </button>

        <Chooser selected={null} onSelect={this.handleDropdownSelect}>
          {this.renderOptions(teams)}
          <Chooser.Option value="" className="btn block hover-bg-silver">
            <div>All teams</div>
          </Chooser.Option>
        </Chooser>
      </Dropdown>
    );
  }

  renderLabel() {
    if (this.props.selected) {
      const selectedTeam = this.teamEdges.filter(Boolean).find(({ node }) => node && node.slug === this.props.selected);
      if (selectedTeam && selectedTeam.node) {
        return <Emojify className="block" text={selectedTeam.node.name} />;
      }
    }
    return "All teams";
  }

  renderOptions(teams) {
    return teams.map((team) =>
      (<Chooser.Option key={team.id} value={team.slug} className="btn block hover-bg-silver line-height-3">
        <Emojify className="block" text={team.name} />
        {team.description && <Emojify className="block dark-gray regular h5 mt1" text={team.description} />}
      </Chooser.Option>)
    );
  }

  handleDropdownSelect = (slug: string) => {
    if (this.dropdownNode) {
      this.dropdownNode.setShowing(false);
      this.props.onTeamChange(slug);
    }
  };
}

export default createFragmentContainer(Teams, graphql`
  fragment Teams_organization on Organization {
    teams(first: 100) {
      edges {
        node {
          id
          name
          slug
          description
          permissions {
            pipelineView {
              allowed
            }
          }
        }
      }
    }
  }
`);
