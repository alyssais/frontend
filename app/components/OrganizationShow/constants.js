// @flow

export const PIPELINES_INITIAL_PAGE_SIZE = 2;
export const PIPELINES_PAGE_SIZE = 2;
