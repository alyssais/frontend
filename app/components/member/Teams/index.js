import React from 'react';
import PropTypes from 'prop-types';
import Relay from 'react-relay/classic';

import Panel from 'app/components/shared/Panel';
import ShowMoreFooter from 'app/components/shared/ShowMoreFooter';

import Row from './row';
import Chooser from './chooser';

const INITIAL_PAGE_SIZE = 5;
const PAGE_SIZE = 20;

class TeamMemberships extends React.PureComponent {
  static displayName = "Member.Teams";

  static propTypes = {
    organizationMember: PropTypes.shape({
      uuid: PropTypes.string.isRequired,
      organization: PropTypes.shape({
        permissions: PropTypes.shape({
          teamAdmin: PropTypes.shape({
            allowed: PropTypes.bool.isRequired
          }).isRequired
        }).isRequired
      }).isRequired,
      teams: PropTypes.shape({
        count: PropTypes.number.isRequired,
        edges: PropTypes.arrayOf(
          PropTypes.shape({
            node: PropTypes.object.isRequired
          }).isRequired
        ).isRequired
      }).isRequired
    }),
    relay: PropTypes.object.isRequired
  };

  state = {
    loading: false
  };

  render() {
    return (
      <div>
        {this.renderTeamAddButton()}
        <Panel>
          {this.renderTeams()}
          <ShowMoreFooter
            connection={this.props.organizationMember.teams}
            label="teams"
            loading={this.state.loading}
            onShowMore={this.handleShowMoreTeams}
          />
        </Panel>
      </div>
    );
  }

  renderTeamAddButton() {
    if (this.props.organizationMember.organization.permissions.teamAdmin.allowed) {
      return (
        <Chooser
          organizationMember={this.props.organizationMember}
          onChoose={this.handleTeamMemberAdd}
        />
      );
    }
  }

  handleTeamMemberAdd = () => {
    this.props.relay.forceFetch();
  };

  renderTeams() {
    const teams = this.props.organizationMember.teams.edges;

    if (!teams.length) {
      return (
        <div className="p3">
          This user is not a member of any teams.
        </div>
      );
    }

    return teams.map(({ node }) => (
      <Row
        key={node.id}
        teamMember={node}
      />
    ));
  }

  handleShowMoreTeams = () => {
    this.setState({ loading: true });

    let { teamsPageSize } = this.props.relay.variables;

    teamsPageSize += PAGE_SIZE;

    this.props.relay.setVariables(
      { teamsPageSize },
      (readyState) => {
        if (readyState.done) {
          this.setState({ loading: false });
        }
      }
    );
  };
}

export default Relay.createContainer(TeamMemberships, {
  initialVariables: {
    teamsPageSize: INITIAL_PAGE_SIZE
  },

  fragments: {
    organizationMember: () => Relay.QL`
      fragment on OrganizationMember {
        uuid
        user {
          id
        }
        organization {
          permissions {
            teamAdmin {
              allowed
            }
          }
        }
        teams(first: $teamsPageSize, order: NAME) {
          ${ShowMoreFooter.getFragment('connection')}
          count
          edges {
            node {
              id
              ${Row.getFragment('teamMember')}
            }
          }
        }
        ${Chooser.getFragment('organizationMember')}
      }
    `
  }
});
