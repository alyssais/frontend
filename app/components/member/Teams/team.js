import React from 'react';
import PropTypes from 'prop-types';
import Relay from 'react-relay/compat';

import TeamLabels from 'app/components/team/Labels';
import Emojify from 'app/components/shared/Emojify';

class Team extends React.PureComponent {
  static displayName = "Member.Edit.TeamMemberships.Chooser.Team";

  static propTypes = {
    team: PropTypes.shape({
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired
    }).isRequired
  };

  render() {
    return (
      <div>
        <div className="flex items-center">
          <Emojify text={this.props.team.name} className="semi-bold truncate" />
          <TeamLabels team={this.props.team} />
        </div>
        <div className="m0 p0 dark-gray truncate"><Emojify text={this.props.team.description || "n/a"} /></div>
      </div>
    );
  }
}

export default Relay.createContainer(Team, {
  fragments: {
    team: () => Relay.QL`
      fragment on Team {
        name
        description
        ${TeamLabels.getFragment('team')}
      }
    `
  }
});
