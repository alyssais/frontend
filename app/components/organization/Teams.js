/* @flow */
import React from 'react';
import Relay from 'react-relay/classic';

import Dropdown from 'app/components/shared/Dropdown';
import Chooser from 'app/components/shared/Chooser';
import Emojify from 'app/components/shared/Emojify';
import Icon from 'app/components/shared/Icon';

type Props = {
  selected?: string,
  organization: Object,
  onTeamChange: Function
};

class Teams extends React.Component {
  props: Props;

  render() {
    // Collect all the teams that we're allowed to see pipelines on
    const teams = [];
    for (const edge of this.props.organization.teams.edges) {
      if (edge.node.permissions.pipelineView.allowed) {
        teams.push(edge.node);
      }
    }

    // Don't render the select if there aren't any teams that can be viewed
    if (teams.length === 0) {
      return null;
    }

    return (
      <Dropdown className="ml4" width={300} ref={(dropdownNode) => this.dropdownNode = dropdownNode}>
        <button className="h3 px0 py1 m0 light dark-gray inline-block btn" style={{ fontSize: 16 }}>
          <div className="flex">
            <span className="flex items-center">
              <span className="truncate">
                {this.renderLabel()}
              </span>
            </span>
            <span className="flex items-center">
              <Icon icon="down-triangle" style={{ width: 8, height: 8, marginLeft: '.5em' }} />
            </span>
          </div>
        </button>

        <Chooser selected={null} onSelect={this.handleDropdownSelect}>
          {this.renderOptions(teams)}
          <Chooser.Option value="" className="btn block hover-bg-silver">
            <div>All teams</div>
          </Chooser.Option>
        </Chooser>
      </Dropdown>
    );
  }

  renderLabel() {
    if (this.props.selected) {
      for (const edge of this.props.organization.teams.edges) {
        if (edge.node.slug === this.props.selected) {
          return (
            <Emojify className="block" text={edge.node.name} />
          );
        }
      }
    }

    return "All teams";
  }

  renderOptions(teams) {
    return teams.map((team) =>
      (<Chooser.Option key={team.id} value={team.slug} className="btn block hover-bg-silver line-height-3">
        <Emojify className="block" text={team.name} />
        {team.description && <Emojify className="block dark-gray regular h5 mt1" text={team.description} />}
      </Chooser.Option>)
    );
  }

  handleDropdownSelect = (slug) => {
    this.dropdownNode.setShowing(false);
    this.props.onTeamChange(slug);
  };
}

export default Relay.createContainer(Teams, {
  fragments: {
    organization: () => Relay.QL`
      fragment on Organization {
        teams(first: 100) {
          edges {
            node {
              id
              name
              slug
              description
              permissions {
                pipelineView {
                  allowed
                }
              }
            }
          }
        }
      }
    `
  }
});
