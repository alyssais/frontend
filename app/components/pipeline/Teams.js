import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Chooser from 'app/components/shared/Chooser';
import Emojify from 'app/components/shared/Emojify';

const PipelineChooser = styled(Chooser)`
  margin: -5px;
`;

class PipelineTeams extends React.Component {
  static propTypes = {
    organization: PropTypes.object.isRequired,
    selected: PropTypes.array.isRequired,
    onTeamSelect: PropTypes.func.isRequired,
    onTeamDeselect: PropTypes.func.isRequired
  };

  state = {
    selected: this.props.selected
  };

  render() {
    return (
      <PipelineChooser selected={this.state.selected} multiple={true} onSelect={this.handleTeamSelect}>
        {
          this.props.organization.teams.edges.map((edge) => {
            return (
              <Chooser.Option
                tag="button"
                key={edge.node.id}
                value={edge.node.id}
                data={{ team: edge.node }}
                className="btn border border-gray rounded m1 regular user-select-none"
                selectedClassName="border-lime"
              >
                <Emojify text={edge.node.name} />
              </Chooser.Option>
            );
          })
        }
      </PipelineChooser>
    );
  }

  handleTeamSelect = (id, data) => {
    const selected = this.state.selected;

    // Toggle its selected state and fire callbacks
    const index = this.state.selected.indexOf(id);
    if (index !== -1) {
      selected.splice(index, 1);
      this.props.onTeamDeselect(data.team);
    } else {
      selected.push(id);
      this.props.onTeamSelect(data.team);
    }

    this.setState({ selected: selected });
  };
}

export default PipelineTeams;
