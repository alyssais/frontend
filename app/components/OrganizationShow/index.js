// @flow

import * as React from 'react';
import { QueryRenderer, graphql } from 'react-relay/compat';
import PropTypes from 'prop-types';
import DocumentTitle from 'react-document-title';
import styled from 'styled-components';
import { stringify } from 'query-string';
import Button from 'app/components/shared/Button';
import Icon from 'app/components/shared/Icon';
import SectionLoader from 'app/components/shared/SectionLoader';
import PageWithContainer from 'app/components/shared/PageWithContainer';
import SearchField from 'app/components/shared/SearchField';
import Pipelines from './Pipelines';
import Teams from './Teams';
import RelayModernPreloader from 'app/lib/RelayModernPreloader';
import Environment from 'app/lib/relay/environment';
import * as constants from './constants';

import type { OrganizationShowQueryResponse } from './__generated__/OrganizationShowQuery.graphql';
type Organization = $NonMaybeType<$ElementType<OrganizationShowQueryResponse, 'organization'>>;

function canCreatePipelineForOrganization(organization: Organization): boolean {
  return (
    organization &&
    organization.permissions &&
    organization.permissions.pipelineCreate &&
    organization.permissions.pipelineCreate.code === "not_member_of_team"
  ) ? false : true;
}

const FilterField = styled(SearchField)`
  flex-basis: 100%;
  order: 3;
  color: gray;

  & > input {
    color: inherit;
    font-size: inherit;
    font-weight: inherit;
  }

  @media (min-width: 480px) {
    font-size: 16px;
    flex-basis: 200px;
    /*
    NOTE: the large-screen margin-left is ~half right to adjust for
    the position of SearchField's icon
    */
    margin-left: .5em;
    margin-right: 1em;
    margin-top: -.25em;
    order: initial;
  }
`;

FilterField.defaultProps = {
  className: 'light flex-auto'
};

type State = {
  pageSize: number,
  pipelineFilter: ?string,
  teamFilter: ?string
};

type Props = {
  params: {
    organization: string
  },
  location: {
    query: {
      team?: string,
      filter?: string
    }
  }
};

export default class OrganizationShow extends React.Component<Props, State> {
  environment = Environment.get();
  state = {
    pageSize: constants.PIPELINES_INITIAL_PAGE_SIZE,
    teamFilter: this.teamFilter,
    pipelineFilter: this.nameFilter
  };

  static query = graphql`
    query OrganizationShowQuery(
      $organizationSlug: ID!
      $teamSearch: TeamSelector
      $pageSize: Int!
      $pipelineFilter: String
    ) {
      organization(slug: $organizationSlug) {
        ...Teams_organization
        ...Pipelines_organization @arguments(
          teamSearch: $teamSearch
          pageSize: $pageSize
          pipelineFilter: $pipelineFilter
        )
        id
        slug
        name
        permissions {
          pipelineCreate {
            code
            allowed
            message
          }
        }
      }
    }
  `;

  static contextTypes = {
    router: PropTypes.object.isRequired
  };

  get organizationSlug(): string {
    return this.props.params.organization;
  }

  get teamFilter(): ?string {
    return this.props.location.query.team || null;
  }

  get nameFilter(): ?string {
    return this.props.location.query.filter || null;
  }

  get queryVariables() {
    return {
      organizationSlug: this.organizationSlug,
      teamSearch: this.state.teamFilter,
      pipelineFilter: this.state.pipelineFilter,
      pageSize: this.state.pageSize
    };
  }

  constructor(props: Props) {
    super(props);

    RelayModernPreloader.preload(
      OrganizationShow.query,
      this.queryVariables,
      this.environment
    );
  }

  render() {
    return (
      <QueryRenderer
        dataFrom="STORE_THEN_NETWORK"
        environment={this.environment}
        query={OrganizationShow.query}
        variables={this.queryVariables}
        render={this.renderQuery()}
      />
    );
  }

  /* eslint-disable react/no-unused-prop-types */
  renderQuery = () => ({ error, props }: { error: ?Error, props: OrganizationShowQueryResponse }) => {
    if (error) {
      return (<div>BONK!</div>);
    }

    if (!props) {
      return (<SectionLoader />);
    }

    if (props && props.organization) {
      const { organization } = props;

      return (
        <DocumentTitle title={`${organization.name}`}>
          <div>
            <PageWithContainer>
              <div className="flex flex-wrap items-start mb2">
                <h1 className="h1 p0 m0 regular line-height-1 inline-block">Pipelines</h1>
                <Teams
                  selected={this.teamFilter}
                  organization={organization}
                  onTeamChange={this.handleTeamChange(organization)}
                />
                <FilterField
                  borderless={true}
                  onChange={this.handleFilterChange(organization)}
                  defaultValue={this.nameFilter}
                  searching={false}
                  placeholder="Filter"
                  autofocus={true}
                />
                {this.renderNewPipelineButton(organization)}
              </div>
              <Pipelines
                organization={props.organization}
                teamFilter={this.teamFilter}
                nameFilter={this.nameFilter}
              />
            </PageWithContainer>
          </div>
        </DocumentTitle>
      );
    }
  }

  renderNewPipelineButton(organization: Organization) {
    // Don't render the "New Pipeline" button if they're not allowed to due to
    // a `not_member_of_team` permsission error.
    if (!canCreatePipelineForOrganization(organization)) {
      return null;
    }

    // Attach the current team to the "New Pipeline" URL
    let newPipelineURL = `/organizations/${organization.slug}/pipelines/new`;
    if (this.teamFilter) {
      newPipelineURL += `?team=${this.teamFilter}`;
    }

    return (
      <Button
        theme="default"
        outline={true}
        className="p0 ml-auto flex circle items-center justify-center"
        style={{ width: 34, height: 34 }}
        href={newPipelineURL}
        title="New Pipeline"
      >
        <Icon icon="plus" title="New Pipeline" />
      </Button>
    );
  }

  handleTeamChange = (organization: Organization) => (team: string) => {
    this.updateRoute(organization, { team });
  };

  handleFilterChange = (organization: Organization) => (filter: string) => {
    this.updateRoute(organization, { filter });
  };

  updateRoute = (organization: Organization, params: {|filter?: string, team?: string|}) => {
    const query = stringify(
      Object
        .entries({ ...this.props.location.query, ...params })
        .reduce((query, [key, value]) => (value ? { ...query, [key]: value } : query), {})
    );

    if (query) {
      this.context.router.push(`/__sneaky__/${organization.slug}?${query}`);
    } else {
      this.context.router.push(`/__sneaky__/${organization.slug}`);
    }
  };
}
