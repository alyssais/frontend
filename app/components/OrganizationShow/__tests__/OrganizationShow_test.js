/* eslint-disable */

import * as React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import Environment from 'app/lib/relay/environment';
import {MockFetch} from 'app/lib/relay/makeFetch';
import OrganizationShow from '..';
import Teams from '../Teams';

import { graphql, buildClientSchema } from 'graphql';
import { addMockFunctionsToSchema } from 'graphql-tools';
import * as introspectionResult from 'app/graph/schema.json';

async function Derp() {
  const schema = buildClientSchema(introspectionResult.data);
  addMockFunctionsToSchema({schema, mocks: {
    DateTime: () => '123123123123',
    Pipeline: () => ({
      name: 'Foobar',
      description: 'OMGOMGOMGOMGOMGOGMGMG,'
    })
  }});

  const query = OrganizationShow.query.modern().text;
  const foo = await graphql(schema, query, null, null, {
    organizationSlug: 'foo',
    pageSize: 30
  });

  return foo;
}

jest.mock('app/lib/relay/makeFetch')

class MockRouterContext extends React.Component {
  static childContextTypes = {
    router: PropTypes.object,
  }

  getChildContext = () => ({
    router: {
      test: 2,
    },
  })

  render() {
    return this.props.children;
  }
}

describe.only('OrganizationShow', () => {
  test('it renders', async (done) => {

    const blah = Derp();

    Environment.create();
    MockFetch.mockResolvedValue(blah)

    const render = renderer.create(
      <MockRouterContext>
        <OrganizationShow
          {...{
            params: {
              organization: 'test'
            },
            location: {
              query: {}
            }
          }}
        />
      </MockRouterContext>
    );

    setTimeout(() => {
      expect(render.toJSON()).toMatchSnapshot()
      done()
    }, 2000)

    // console.log('----------------------------')
    // console.log(render.toTree())
    // console.log(render.toJSON())
    // console.log(render.root.children)
    // console.log('----------------------------')
    // console.log(MockFetch)
  });
});
