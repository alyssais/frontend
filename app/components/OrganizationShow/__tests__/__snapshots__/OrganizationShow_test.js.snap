// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationShow it renders 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="flex flex-wrap items-start mb2"
    >
      <h1
        className="h1 p0 m0 regular line-height-1 inline-block"
      >
        Pipelines
      </h1>
      <span
        className="ml4"
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <button
          className="h3 px0 py1 m0 light dark-gray inline-block btn"
          style={
            Object {
              "fontSize": 16,
            }
          }
        >
          <div
            className="flex"
          >
            <span
              className="flex items-center"
            >
              <span
                className="truncate"
              >
                All teams
              </span>
            </span>
            <span
              className="flex items-center"
            >
              <svg
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                height="20px"
                style={
                  Object {
                    "fill": "currentColor",
                    "height": 8,
                    "marginLeft": ".5em",
                    "verticalAlign": "middle",
                    "width": 8,
                  }
                }
                viewBox="0 0 20 20"
                width="20px"
              />
            </span>
          </div>
        </button>
      </span>
      <div
        className="relative dark-gray light flex-auto sc-EHOje hAGMXy"
      >
        <svg
          className="absolute pointer-events-none"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          height="20px"
          style={
            Object {
              "color": "currentColor",
              "fill": "currentColor",
              "height": "1em",
              "left": ".75em",
              "top": ".68em",
              "verticalAlign": "middle",
              "width": "1em",
            }
          }
          viewBox="0 0 20 20"
          width="20px"
        />
        <input
          className="input borderless"
          onChange={[Function]}
          placeholder="Filter"
          style={
            Object {
              "margin": 0,
              "paddingLeft": "2em",
            }
          }
          type="text"
          value=""
        />
      </div>
      <a
        className="btn nowrap p0 ml-auto flex circle items-center justify-center btn-outline border-gray hover-black"
        href="/organizations/Hello World/pipelines/new"
        style={
          Object {
            "height": 34,
            "width": 34,
          }
        }
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          height="20px"
          style={
            Object {
              "fill": "currentColor",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 20 20"
          width="20px"
        />
      </a>
    </div>
    <div>
      <div
        className="flex items-stretch border border-gray rounded px2 mb2"
        style={
          Object {
            "height": 82,
          }
        }
      >
        <div
          className="flex flex-none items-center pl2 pr2"
        >
          <svg
            height={32}
            viewBox="0 0 32 32"
            width={32}
          >
            <defs>
              <circle
                cx="16"
                cy="16"
                fill="none"
                id="BuildState_4c03d7c9-4853-4e63-9180-9f834e007ebf_circle"
                r="15"
                stroke="#cdcccc"
                strokeWidth={4}
              />
              <clipPath
                id="BuildState_4c03d7c9-4853-4e63-9180-9f834e007ebf_strokeClipPath"
              >
                <use
                  xlinkHref="#BuildState_4c03d7c9-4853-4e63-9180-9f834e007ebf_circle"
                />
              </clipPath>
            </defs>
            <use
              clipPath="url(#BuildState_4c03d7c9-4853-4e63-9180-9f834e007ebf_strokeClipPath)"
              xlinkHref="#BuildState_4c03d7c9-4853-4e63-9180-9f834e007ebf_circle"
            />
            <path
              d="M11,16H21"
              fill="none"
              stroke="#cdcccc"
              strokeWidth={2}
            />
          </svg>
        </div>
        <a
          className="flex flex-auto items-center px2 text-decoration-none color-inherit mr3"
          href="Hello World"
        >
          <div
            className="truncate"
          >
            <h2
              className="inline h3 regular m0 line-height-2"
            >
              <span>
                Foobar
              </span>
            </h2>
            <div
              className="truncate dark-gray"
              style={
                Object {
                  "marginTop": 3,
                }
              }
            >
              <span
                className="h4 regular"
              >
                OMGOMGOMGOMGOMGOGMGMG,
              </span>
            </div>
          </div>
        </a>
        <div
          className="flex items-center flex-none ml3 xs-hide sm-hide pr4"
        >
          <div
            style={
              Object {
                "width": 239,
              }
            }
          >
            <div
              className="h6 regular dark-gray mb1 line-height-1"
            >
              Hello World
            </div>
            <div
              className="flex"
              style={
                Object {
                  "height": 35,
                }
              }
            >
              <div
                className="bg-gray animation-loading-bar"
                style={
                  Object {
                    "height": 3,
                    "marginTop": "auto",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
        </div>
        <div
          className="flex flex-none items-center justify-start xs-hide ml2"
        >
          <div
            className="flex items-center"
          >
            <a
              className="flex flex-column text-decoration-none color-inherit"
              href="Hello World"
              style={
                Object {
                  "width": "7em",
                }
              }
            >
              <span
                className="h6 regular dark-gray truncate"
              >
                4cd0f5c3-14fa-4ac7-8973-920e2b93a3c7
              </span>
              <span
                className="h1 light m0 line-height-1 truncate"
              >
                Hello World
              </span>
            </a>
            <a
              className="flex flex-column text-decoration-none color-inherit"
              href="Hello World"
              style={
                Object {
                  "width": "7em",
                }
              }
            >
              <span
                className="h6 regular dark-gray truncate"
              >
                7f84167a-456b-4633-ba3d-b1297ff506d1
              </span>
              <span
                className="h1 light m0 line-height-1 truncate"
              >
                Hello World
              </span>
            </a>
          </div>
        </div>
        <div
          className="flex flex-none flex-column justify-center ml-auto pr2"
        >
          <button
            className="btn p0 dark-gray line-height-1"
            onClick={[Function]}
          >
            <svg
              className="sc-bdVaJa daifPi"
              height="15px"
              viewBox="0 0 16 15"
              width="20px"
            >
              <title>
                Favorite
              </title>
              <g
                className="star"
                transform="translate(-1103, -19)"
              >
                <polygon
                  points="1111 31 1106.29772 33.472136 1107.19577 28.236068 1103.39155 24.527864 1108.64886 23.763932 1111 19 1113.35114 23.763932 1118.60845 24.527864 1114.80423 28.236068 1115.70228 33.472136"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
      <hr
        className="my4 bg-gray mx-auto max-width-1 border-none height-0"
        style={
          Object {
            "height": 1,
          }
        }
      />
      <div
        className="flex items-stretch border border-gray rounded px2 mb2"
        style={
          Object {
            "height": 82,
          }
        }
      >
        <div
          className="flex flex-none items-center pl2 pr2"
        >
          <svg
            height={32}
            viewBox="0 0 32 32"
            width={32}
          >
            <defs>
              <circle
                cx="16"
                cy="16"
                fill="none"
                id="BuildState_7694c2c9-aad9-4b8e-9206-b9ae21247a3f_circle"
                r="15"
                stroke="#cdcccc"
                strokeWidth={4}
              />
              <clipPath
                id="BuildState_7694c2c9-aad9-4b8e-9206-b9ae21247a3f_strokeClipPath"
              >
                <use
                  xlinkHref="#BuildState_7694c2c9-aad9-4b8e-9206-b9ae21247a3f_circle"
                />
              </clipPath>
            </defs>
            <use
              clipPath="url(#BuildState_7694c2c9-aad9-4b8e-9206-b9ae21247a3f_strokeClipPath)"
              xlinkHref="#BuildState_7694c2c9-aad9-4b8e-9206-b9ae21247a3f_circle"
            />
            <path
              d="M11,16H21"
              fill="none"
              stroke="#cdcccc"
              strokeWidth={2}
            />
          </svg>
        </div>
        <a
          className="flex flex-auto items-center px2 text-decoration-none color-inherit mr3"
          href="Hello World"
        >
          <div
            className="truncate"
          >
            <h2
              className="inline h3 regular m0 line-height-2"
            >
              <span>
                Foobar
              </span>
            </h2>
            <div
              className="truncate dark-gray"
              style={
                Object {
                  "marginTop": 3,
                }
              }
            >
              <span
                className="h4 regular"
              >
                OMGOMGOMGOMGOMGOGMGMG,
              </span>
            </div>
          </div>
        </a>
        <div
          className="flex items-center flex-none ml3 xs-hide sm-hide pr4"
        >
          <div
            style={
              Object {
                "width": 239,
              }
            }
          >
            <div
              className="h6 regular dark-gray mb1 line-height-1"
            >
              Hello World
            </div>
            <div
              className="flex"
              style={
                Object {
                  "height": 35,
                }
              }
            >
              <div
                className="bg-gray animation-loading-bar"
                style={
                  Object {
                    "height": 3,
                    "marginTop": "auto",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
        </div>
        <div
          className="flex flex-none items-center justify-start xs-hide ml2"
        >
          <div
            className="flex items-center"
          >
            <a
              className="flex flex-column text-decoration-none color-inherit"
              href="Hello World"
              style={
                Object {
                  "width": "7em",
                }
              }
            >
              <span
                className="h6 regular dark-gray truncate"
              >
                57b36496-afbe-4df7-af76-57493e24dd37
              </span>
              <span
                className="h1 light m0 line-height-1 truncate"
              >
                Hello World
              </span>
            </a>
            <a
              className="flex flex-column text-decoration-none color-inherit"
              href="Hello World"
              style={
                Object {
                  "width": "7em",
                }
              }
            >
              <span
                className="h6 regular dark-gray truncate"
              >
                c1ab5151-e3e2-4d97-983a-90df841ee7f4
              </span>
              <span
                className="h1 light m0 line-height-1 truncate"
              >
                Hello World
              </span>
            </a>
          </div>
        </div>
        <div
          className="flex flex-none flex-column justify-center ml-auto pr2"
        >
          <button
            className="btn p0 dark-gray line-height-1"
            onClick={[Function]}
          >
            <svg
              className="sc-bdVaJa ixRRAY"
              height="15px"
              viewBox="0 0 16 15"
              width="20px"
            >
              <title>
                Favorite
              </title>
              <g
                className="star"
                transform="translate(-1103, -19)"
              >
                <polygon
                  points="1111 31 1106.29772 33.472136 1107.19577 28.236068 1103.39155 24.527864 1108.64886 23.763932 1111 19 1113.35114 23.763932 1118.60845 24.527864 1114.80423 28.236068 1115.70228 33.472136"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
